File: typescript/2024/day5.ts
import * as R from 'ramda';

import { parseInt10 } from '../lib/fp';

const content = (await Bun.file('../inputs/2024/Day5/input.txt').text()).trim();

const [a, b] = content.split('\n\n').map(section => section.split('\n'));

const rules = a.map(line => line.split('|').map(parseInt10)) as [number, number][];
const pages = b.map(line => line.split(',').map(parseInt10));

const inRightOrder = (rule: [number, number], page: number[]) => {
  const [i1, i2] = rule.map(v => page.indexOf(v)) as [number, number];
  if (i1 === -1 || i2 === -1) return true;
  return i1 < i2;
};

const getMiddleValue = <T>(arr: T[]): T => {
  const i = Math.floor(arr.length / 2);
  return arr[i];
};

const [correctOrder, incorrectOrder] = R.partition(page => rules.every(rule => inRightOrder(rule, page)), pages);

const r1 = R.sum(correctOrder.map(getMiddleValue));
console.log(r1);

const updated = incorrectOrder.map(line =>
  R.sort((l, r) => (rules.find(([ll, rr]) => r === rr && l === ll) ? -1 : 1), line),
);

const r2 = R.sum(updated.map(getMiddleValue));
console.log(r2);

File: haskell/2024/Day5.hs
module Day5 where

import Data.List
import Data.List.Split
import Data.Ord
import Data.Tuple.Common

parseInput :: [String] -> ([(Int, Int)], [[Int]])
parseInput xs =
  let (a, b) = toTuple $ splitOn [""] xs
      rules = map (toTuple . map read . splitOn "|") a
      pages = map (map read . splitOn ",") b
   in (rules, pages)

inRightOrder :: (Int, Int) -> [Int] -> Bool
inRightOrder rule page =
  let indexes = tmap (`elemIndex` page) rule
      r = sequenceT indexes
   in case r of
        Nothing -> True
        Just (a, b) -> a < b

getMiddleValue :: [a] -> a
getMiddleValue xs = xs !! (length xs `div` 2)

main' :: IO ()
main' = do
  (rules, pages) <- parseInput . lines <$> readFile "../inputs/2024/Day5/input.txt"
  -- print rules
  -- print pages
  let (correctOrder, incorrectOrder) = partition (\page -> all (`inRightOrder` page) rules) pages
  -- print correctOrder
  let r1 = sum $ map getMiddleValue correctOrder
  print r1
  let updated = map (sortBy (\l r -> if any (\(ll, rr) -> ll == l && rr == r) rules then LT else GT)) incorrectOrder
  let r2 = sum $ map getMiddleValue updated
  print r2
